compiler/src/dotty/tools/dotc/typer/Namer.scala

```diff
@@ -902,9 +906,9 @@ class Namer { typer: Typer =>
             && (definesMember || inheritsConcreteMember)
           )
           ||
-          // remove synthetic constructor of a java Record if it clashes with a non-synthetic constructor
-          (denot.isConstructor
-            && denot.owner.is(JavaDefined) && denot.owner.derivesFrom(defn.JavaRecordClass)
+          // remove synthetic constructor or method of a java Record if it clashes with a non-synthetic constructor
+          (isJavaRecord(denot.owner)
+            && (denot.isConstructor || definesMember)
             && denot.owner.unforcedDecls.lookupAll(denot.name).exists(c => c != denot.symbol && c.info.matches(denot.info))
```

Comment:
Is `denot.owner.unforcedDecls.lookupAll(denot.name).exists(c => c != denot.symbol && c.info.matches(denot.info))` redundant with `definesMember`?

Reply:
Yes I think so, but I think that we ought to switch it out since we want the `denot.info` to match exactly in this case. I will change the logic here.

Reply:
@noti0na1 Changed the logic and the tests passed locally.

Reply:
Great! I'm not that familiar with other parts, let's wait for review from Nicolas.